aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -1027,21 +1027,35 @@ Tensor trace_cpu(const Tensor& self) {
   return result;
 }
 
-Tensor prod(const Tensor& self, int64_t dim, bool keepdim, c10::optional<ScalarType> opt_dtype) {
-  ScalarType dtype = get_dtype_from_self(self, opt_dtype, true);
-  Tensor result = create_reduction_result(self, dim, keepdim, dtype);
-  native::prod_out_impl(result, self, dim, keepdim, dtype);
-  return result;
+void impl_func_prod(
+    const Tensor& self,
+    IntArrayRef dims,
+    bool keepdim,
+    c10::optional<ScalarType> dtype,
+    const Tensor& result) {
+  auto iter = meta::make_reduction_from_out_ty(self, result, dims, keepdim, result.scalar_type());
+  if (iter.numel() == 0) {
+    result.fill_(1);
+  } else {
+    prod_stub(iter.device_type(), iter);
+  }
 }
 
-Tensor prod(const Tensor &self, c10::optional<ScalarType> opt_dtype) {
-  ScalarType dtype = get_dtype_from_self(self, opt_dtype, true);
-  Tensor result = create_reduction_result(self, {}, false, dtype);
-  return at::native::prod_out_impl(result, self, {}, false, dtype);
+TORCH_IMPL_FUNC(prod_out)
+(const Tensor& self,
+ int64_t dim,
+ bool keepdim,
+ c10::optional<ScalarType> dtype,
+ const Tensor& result) {
+  impl_func_prod(self, dim, keepdim, dtype, result);
 }
 
-Tensor& prod_out(const Tensor& self, int64_t dim, bool keepdim, c10::optional<ScalarType> dtype, Tensor& result) {
-  return at::native::prod_out_impl(result, self, dim, keepdim, dtype);
+Tensor prod(const Tensor &self, c10::optional<ScalarType> opt_dtype) {
+  auto dtype = get_dtype_from_self(self, opt_dtype, true);
+  auto shape = meta::get_reduction_shape(self, {}, false);
+  Tensor result = at::empty(shape, self.options().dtype(dtype));
+  impl_func_prod(self, {}, false, dtype, result);
+  return result;
```

Comment:
This looks suspicious. How come we have to define `prod` here?

Reply:
The problem is that we can't call `prod.dim_int` since it expects only one dimension to reduce (`prod`reduces all of them). This differs from `sum.dim_IntList`, that expects many dimensions: `int[1]` (i.e. we can call `sum.dim_IntList` inside `sum` kernel)